ops/relation.py

```diff
@@ -434,6 +434,12 @@ def _validate_provider(self, event):
         logger.debug("Validating provider(s) : %s", event)
         consumed = self.consumes
 
+        if self.name not in self.framework.model.relations:
+            logger.error("No %s relation established, cannot validate "
+                         "provider", self.name)
+            event.defer()
+            return
+
         for relation in self.framework.model.relations[self.name]:
```

Comment:
```suggestion
        for relation in self.framework.model.relations.get(self.name, []):
```
because there is no guarantee that the relation would ever be established (e.g. karma never gets related to alertmanager), so no point to error or defer.

Reply:
This is a misunderstanding of how this object is populated. Think of it this way:

When you deploy your charm initially, without relations established, do you see this exception?

It is not populated from relations being added. Instead, [`metadata.yaml` is ingested](https://github.com/canonical/operator/blob/b1a8fffc053fbbdaf43b826e78c220bfdae4c8fe/ops/main.py#L357) and [passed to `CharmMeta`](https://github.com/canonical/operator/blob/b1a8fffc053fbbdaf43b826e78c220bfdae4c8fe/ops/charm.py#L722). This is then used [create a `RelationMapping`](https://github.com/canonical/operator/blob/b1a8fffc053fbbdaf43b826e78c220bfdae4c8fe/ops/model.py#L63) in the model,  where this is queried from. (This is also used to [create all of the events](https://github.com/canonical/operator/blob/b1a8fffc053fbbdaf43b826e78c220bfdae4c8fe/ops/charm.py#L603) that your charm uses).

It does not matter whether a relation is established in order for this key to exist. I suspect this is actually a bug in populating these objects during upgrades, but deferring is a confirmed workaround which you can use in your charm for now.

Reply:
I see. But `_validate_provider` is called only once - on charm upgrade. That is why it is not seen on startup.